core/src/main/scala/com/avsystem/commons/di/Component.scala

```diff
@@ -171,10 +171,15 @@ final class Component[+T](
           val resultFuture =
             Future.traverse(dependencies)(_.doInit(starting = false))
               .flatMap(resolvedDeps => creator(resolvedDeps)(ec))
+              .flatMap {
+                case component: AsyncInitializingComponent => component.init().map(_ => component)
+                case component: InitializingComponent => Future(component.init()).map(_ => component)
+                case component => Future.successful(component)
+              }
               .recoverNow {
                 case NonFatal(cause) =>
                   throw ComponentInitializationException(this, cause)
-              }
+              }.asInstanceOf[Future[T]]
```

Comment:
A simple `.flatMap { value => value match {` with mappings to `value` allows to avoid casting here.